comm/sync.go

```
@@ -57,66 +62,81 @@ func fetchBlocks(ctx context.Context, peer *Peer, fromBlockNum uint32, fetched c
 			return nil
 		}
 
-		blocks := make([]*block.Block, 0, len(result))
-		for _, raw := range result {
-			var blk block.Block
-			if err := rlp.DecodeBytes(raw, &blk); err != nil {
-				return errors.Wrap(err, "invalid block")
-			}
-			if blk.Header().Number() != fromBlockNum {
-				return errors.New("broken sequence")
-			}
-			fromBlockNum++
-			blocks = append(blocks, &blk)
+		batch := rawBlockBatch{
+			rawBlocks: result,
+			startNum:  fromBlockNum,
 		}
 
 		select {
-		case fetched <- blocks:
+		case rawBatches <- batch:
 		case <-ctx.Done():
 			return ctx.Err()
 		}
+
+		fromBlockNum += uint32(len(result))
 	}
 }
 
-func warmupBlocks(ctx context.Context, fetched <-chan []*block.Block, warmedUp chan<- *block.Block) {
+func decodeAndWarmupBatches(ctx context.Context, rawBatches <-chan rawBlockBatch, warmedUp chan<- *block.Block) error {
+	errorChan := make(chan error, 1)
+
 	<-co.Parallel(func(queue chan<- func()) {
-		for blocks := range fetched {
-			for _, blk := range blocks {
-				queue <- func() {
-					blk.Header().ID()
-					blk.Header().Beta()
-				}
-				for _, tx := range blk.Transactions() {
-					queue <- func() {
-						tx.ID()
-						tx.UnprovedWork()
-						_, _ = tx.IntrinsicGas()
-						_, _ = tx.Delegator()
+		for batch := range rawBatches {
+			queue <- func() {
+				if err := decodeAndWarmupBatch(ctx, batch, warmedUp); err != nil {
+					select {
+					case errorChan <- err:
```

It looks like errors might be lost if multiple errors occur simultaneously because the error checking happens after co.Parallel() completes right ?

---

Correct, fixed now